### ISSUE TITLE
autoSync seems to be sending to backend but .on('http', callback) isn't called, testing on iOS simulator

### ISSUE THREAD

**Issue by pencilcheck** [transistorsoft/react-native-background-geolocation#383]
## Your Environment
* Plugin version: 2.10.1
* Platform: iOS
* OS version: iOS 11.2
* Device manufacturer / model: iPhone 6s
* React Native version (`react-native -v`): 
react-native-cli: 2.0.1
react-native: 0.48.4

* Plugin config

Printed in the log below

## Expected Behavior

Besides event 'location' callback being called, there should be request to backend from setting url attribute from onHttp callbacks

## Actual Behavior

Only event 'location' callback being called, no request to server, and no callback response from onHttp()

## Steps to Reproduce

1. Launch xcode and ios and app on it
2. Turn on simulation location service in xcode to test location automatically
3. Nothing in the console nor in the server log
4.

## Context

Testing if I can use plugin's HTTP to persist location to backend

## Debug logs

```
══════════════════
║ -[TSLocationManager locationManager:didUpdateLocations:] Enabled: 1 | isMoving: 1 | df: 10.0m | age: 0.0s
╚═══════════════════════════════════════════════════════════
🔵-[TSLocationManager calculateMedianLocationAccuracy:] Median location accuracy: 5.0
🔵-[Settings incrementOdometer:] 65702801.4
ℹ️-[TSLocationManager on:success:failure:] location
ℹ️-[TSLocationManager on:success:failure:] providerchange
ℹ️-[TSLocationManager on:success:failure:] http
⚠️-[Settings load:] Unknown @config: foregroundService
⚠️-[Settings load:] Unknown @config: notificationPriority

╔═══════════════════════════════════════════════════════════
║ -[Settings load:] 
{
    activityRecognitionInterval = 10000;
    activityType = Other;
    autoSync = 1;
    autoSyncThreshold = 1;
    batchSync = 0;
    debug = 0;
    desiredAccuracy = 100;
    desiredOdometerAccuracy = 100;
    disableElasticity = 0;
    disableMotionActivityUpdates = 0;
    disableStopDetection = 0;
    distanceFilter = 10;
    elasticityMultiplier = 1;
    enabled = 1;
    extras =     {
        input =         {
            ping =             {
                accountId = "0ee2850c-dc12-11e7-823e-13566f91a3b1";
                timezone = "America/Los_Angeles";
            };
        };
    };
    geofenceInitialTriggerEntry = 1;
    geofenceProximityRadius = 1000;
    geofenceTemplate = "";
    headers =     {
        Accept = "application/json";
        Authorization = "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidGF4ZGF5X3VzZXIiLCJhdXRoX2FjY291bnRfaWQiOiIwZWUyYTNiNi1kYzEyLTExZTctODIzZS02ZmFkZWE2ZjgxZTMiLCJleHAiOjE1NDc2MjE4NzQsImlhdCI6MTUxNjA4NTg3NCwiYXVkIjoicG9zdGdyYXBocWwiLCJpc3MiOiJwb3N0Z3JhcGhxbCJ9.IbtM83SZqAd-XCrrt4vj4XNwyY9lbHR8G7Ksmbm0D28";
        "Content-Type" = "application/json";
    };
    heartbeatInterval = 60;
    httpRootProperty = variables;
    httpTimeout = 60000;
    isHeartbeatEnabled = 0;
    isMoving = 1;
    locationAuthorizationAlert =     {
        cancelButton = Cancel;
        instructions = "You must enable 'Always' for location services in the device settings";
        settingsButton = Settings;
        titleWhenNotEnabled = "Location tracking is not enabled";
        titleWhenOff = "Location tracking is OFF";
    };
    locationAuthorizationRequest = Always;
    locationTemplate = "{\n    \"input\": {\n      \"ping\": {\n        \"timestamp\": \"<%= timestamp %>\",\n        \"location\": \"(<%= latitude %>, <%= longitude %>)\"\n      }\n    }\n  }";
    locationTimeout = 60;
    locationsOrderDirection = ASC;
    logLevel = 5;
    logMaxDays = 7;
    maxBatchSize = "-1";
    maxDaysToPersist = 1;
    maxRecordsToPersist = 0;
    method = POST;
    minimumActivityRecognitionConfidence = 70;
    odometer = "65702801.38368355";
    params =     {
        query = "\n      mutation CreatePing($input: CreatePingInput!) {\n        createPing(input: $input) {\n          ping {\n            accountId\n            nodeId\n          }\n        }\n      }\n    ";
    };
    pausesLocationUpdatesAutomatically = 1;
    preventSuspend = 0;
    schedule =     (
    );
    schedulerEnabled = 0;
    startOnBoot = 1;
    stationaryRadius = 100;
    stopAfterElapsedMinutes = "-1";
    stopDetectionDelay = 0;
    stopOnStationary = 1;
    stopOnTerminate = 0;
    stopTimeout = 10;
    trackingMode = 1;
    url = "http://localhost:5000/graphql";
    useSignificantChangesOnly = 0;
}
╚═══════════════════════════════════════════════════════════
⚠️-[TSLocationManager start] Already started (ignored)

📍<+41.70407238,-74.18763705> +/- 5.00m (speed -1.00 mps / course -1.00) @ 1/15/18, 11:47:37 PM Pacific Standard Time

╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager locationManager:didUpdateLocations:] Enabled: 1 | isMoving: 1 | df: 10.0m | age: 0.0s
╚═══════════════════════════════════════════════════════════
🔵-[TSLocationManager calculateMedianLocationAccuracy:] Median location accuracy: 5.0
🔵-[Settings incrementOdometer:] 65712801.4

📍<+41.63460209,-74.11125156> +/- 5.00m (speed -1.00 mps / course -1.00) @ 1/15/18, 11:47:38 PM Pacific Standard Time

╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager locationManager:didUpdateLocations:] Enabled: 1 | isMoving: 1 | df: 10.0m | age: 0.0s
╚═══════════════════════════════════════════════════════════
🔵-[TSLocationManager calculateMedianLocationAccuracy:] Median location accuracy: 5.0
🔵-[Settings incrementOdometer:] 65722801.4

📍<+41.56232228,-74.03972964> +/- 5.00m (speed -1.00 mps / course -1.00) @ 1/15/18, 11:47:39 PM Pacific Standard Time

╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager locationManager:didUpdateLocations:] Enabled: 1 | isMoving: 1 | df: 10.0m | age: 0.0s
╚═══════════════════════════════════════════════════════════
🔵-[TSLocationManager calculateMedianLocationAccuracy:] Median location accuracy: 5.0
🔵-[Settings incrementOdometer:] 65732801.4

📍<+41.48621928,-73.97570278> +/- 5.00m (speed -1.00 mps / course -1.00) @ 1/15/18, 11:47:40 PM Pacific Standard Time

╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager locationManager:didUpdateLocations:] Enabled: 1 | isMoving: 1 | df: 10.0m | age: 0.0s
╚═══════════════════════════════════════════════════════════
🔵-[TSLocationManager calculateMedianLocationAccuracy:] Median location accuracy: 5.0
🔵-[Settings incrementOdometer:] 65742801.4

```

The rest of the log is basically the same as it repeats itself on every location change. But just no request to server.

**Comment by christocracy** [comment]
well, your `locationTemplate` and `#params` are messed up.

`#params` is supposed to an `Object` not `JSON`.
`#locationTemplate` is a `String`, not `JSON`

I would have expected to see error messages about this, but I see you've configured `maxRecordsToPersist: 0`, which means you're not interested in saving records (and by extension, posting them to your server).

When you set `maxRecordsToPersist > 0`, you'll start seeing error messages about `#params` and `locationTemplate`.

**Comment by pencilcheck** [comment]
`#params` is JSON in the code.
`#locationTemplate` is String in the code.

```
var backgroundGeolocationConfiguration = {
  // Debug
  debug: false,
  logLevel: BackgroundGeolocation.LOG_LEVEL_VERBOSE,

  // Common
  distanceFilter: 10,
  desiredAccuracy: Platform.OS === 'ios' ? BackgroundGeolocation.DESIRED_ACCURACY_LOW : BackgroundGeolocation.DESIRED_ACCURACY_MEDIUM, // low for iOS and medium for Android
  stopOnTerminate: false, // will continue tracking even if app is terminated, ios and android behavior differs, need this flag for Android's forceReload* to be usable and for iOS to continue tracking in background
  startOnBoot: true, // engage background-tracking after the device reboots, but for android it will not launch the javascript app

  // Activity detection
  stopOnStationary: true, // stop GPS on stationary to save battery life
  stopTimeout: 10, // 10mins timeout before turning off GPS after stopping from moving

  // HTTP
  url: Platform.OS === 'ios' ? GRAPHQL_URL : ANDROID_GRAPHQL_URL,
  maxRecordsToPersist: 1,
  httpRootProperty: 'variables',
  params: {
    query: `
      mutation CreatePing($input: CreatePingInput!) {
        createPing(input: $input) {
          ping {
            accountId
            nodeId
          }
        }
      }
    `
  },
  locationTemplate: `{
    "input": {
      "ping": {
        "timestamp": "<%= timestamp %>",
        "location": "(<%= latitude %>, <%= longitude %>)"
      }
    }
  }`,
  extras: {
    "input": {
      "ping": {
        "accountId": "",
        "timezone": "",
      }
    }
  },


  // WARN: we really need to make sure location update is clearly sent to user so we don't have race condition
  // iOS
  stationaryRadius: 100,
  locationAuthorizationAlert: {
    titleWhenNotEnabled: "Location tracking is not enabled",
    titleWhenOff: "Location tracking is OFF",
    instructions: "You must enable 'Always' for location services in the device settings",
    cancelButton: "Cancel",
    settingsButton: "Settings"
  },

  // Android
  foregroundService: true, // making background-geolocation library works like playing music in the background, can still be killed but will be more resilient to killing
  notificationPriority: BackgroundGeolocation.NOTIFICATION_PRIORITY_MIN, // Notification strongly weighted to bottom of list; notification-bar icon hidden
  //locationUpdateInterval: 30000, // with distanceFilter = 0, set the interval = 30s
  //forceReloadOnLocationChange: true, // has to pair with stopOnTerminate: false, this flag have geolocation library as background service to launch the javascript/react-native app on location change if the app was terminated by user. We need this because uploading ping to server is only done in the javascript app, not geolocation service library
  //forceReloadOnBoot: true, // should be a preference for user, this allows the background service of geolocation to launch javascript code on device reboot
}
```

However after uninstalling the app in the simulator before starting now I get more logs but still no console.log for `onHttp` nor `on('http', fn)`

```
╔═══════════════════════════════════════════════════════════
║ -[TSHttpService flush:error:] 
╚═══════════════════════════════════════════════════════════
✅-[BackgroundTaskManager createBackgroundTask] 290
✅-[TSHttpService schedulePost:] LOCKED: 856AC210-4B75-476C-BC7B-B58483C8DACA
🔵-[TSHttpService doPost:] HTTP Response: 200
✅-[TSHttpService post:error:] DESTROY: 856AC210-4B75-476C-BC7B-B58483C8DACA

╔═══════════════════════════════════════════════════════════
║ -[TSHttpService finish:error:] Success: 1
╚═══════════════════════════════════════════════════════════
✅-[BackgroundTaskManager stopBackgroundTask:]_block_invoke 290 OF (
    290
)

📍<+44.04887551,-72.02204780> +/- 5.00m (speed -1.00 mps / course -1.00) @ 1/16/18, 12:15:29 AM Pacific Standard Time

╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager locationManager:didUpdateLocations:] Enabled: 1 | isMoving: 1 | df: 10.0m | age: 0.0s
╚═══════════════════════════════════════════════════════════
🔵-[TSLocationManager calculateMedianLocationAccuracy:] Median location accuracy: 5.0
🔵-[Settings incrementOdometer:] 4455751.8
ℹ️-[LocationDAO shrink:]_block_invoke maxRecordsToPersist: 1
✅-[TSLocationManager persistLocation:]_block_invoke INSERT: CD63C11C-3AEC-4188-82D9-27CE470D22E4

╔═══════════════════════════════════════════════════════════
║ -[TSHttpService flush:error:] 
╚═══════════════════════════════════════════════════════════
✅-[BackgroundTaskManager createBackgroundTask] 291
✅-[TSHttpService schedulePost:] LOCKED: CD63C11C-3AEC-4188-82D9-27CE470D22E4
🔵-[TSHttpService doPost:] HTTP Response: 200
✅-[TSHttpService post:error:] DESTROY: CD63C11C-3AEC-4188-82D9-27CE470D22E4

╔═══════════════════════════════════════════════════════════
║ -[TSHttpService finish:error:] Success: 1
╚═══════════════════════════════════════════════════════════
✅-[BackgroundTaskManager stopBackgroundTask:]_block_invoke 291 OF (
    291
)

📍<+44.03525546,-72.14537752> +/- 5.00m (speed -1.00 mps / course -1.00) @ 1/16/18, 12:15:30 AM Pacific Standard Time

╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager locationManager:didUpdateLocations:] Enabled: 1 | isMoving: 1 | df: 10.0m | age: 0.0s
╚═══════════════════════════════════════════════════════════
🔵-[TSLocationManager calculateMedianLocationAccuracy:] Median location accuracy: 5.0
🔵-[Settings incrementOdometer:] 4465751.8
ℹ️-[LocationDAO shrink:]_block_invoke maxRecordsToPersist: 1
✅-[TSLocationManager persistLocation:]_block_invoke INSERT: 648675BF-6360-46D0-8FE7-34685E2B0ABC

╔═══════════════════════════════════════════════════════════
║ -[TSHttpService flush:error:] 
╚═══════════════════════════════════════════════════════════
✅-[BackgroundTaskManager createBackgroundTask] 292
✅-[TSHttpService schedulePost:] LOCKED: 648675BF-6360-46D0-8FE7-34685E2B0ABC
🔵-[TSHttpService doPost:] HTTP Response: 200
✅-[TSHttpService post:error:] DESTROY: 648675BF-6360-46D0-8FE7-34685E2B0ABC

╔═══════════════════════════════════════════════════════════
║ -[TSHttpService finish:error:] Success: 1
╚═══════════════════════════════════════════════════════════
✅-[BackgroundTaskManager stopBackgroundTask:]_block_invoke 292 OF (
    292
)

📍<+44.02064478,-72.26845993> +/- 5.00m (speed -1.00 mps / course -1.00) @ 1/16/18, 12:15:31 AM Pacific Standard Time

╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager locationManager:didUpdateLocations:] Enabled: 1 | isMoving: 1 | df: 10.0m | age: 0.0s
╚═══════════════════════════════════════════════════════════
🔵-[TSLocationManager calculateMedianLocationAccuracy:] Median location accuracy: 5.0
🔵-[Settings incrementOdometer:] 4475751.8
ℹ️-[LocationDAO shrink:]_block_invoke maxRecordsToPersist: 1
✅-[TSLocationManager persistLocation:]_block_invoke INSERT: 350C3CFD-9B1C-4BF2-B091-AF250B76986E

╔═══════════════════════════════════════════════════════════
║ -[TSHttpService flush:error:] 
╚═══════════════════════════════════════════════════════════
✅-[BackgroundTaskManager createBackgroundTask] 293
✅-[TSHttpService schedulePost:] LOCKED: 350C3CFD-9B1C-4BF2-B091-AF250B76986E
🔵-[TSHttpService doPost:] HTTP Response: 200
✅-[TSHttpService post:error:] DESTROY: 350C3CFD-9B1C-4BF2-B091-AF250B76986E

╔═══════════════════════════════════════════════════════════
║ -[TSHttpService finish:error:] Success: 1
╚═══════════════════════════════════════════════════════════
✅-[BackgroundTaskManager stopBackgroundTask:]_block_invoke 293 OF (
    293
)

📍<+44.00458351,-72.39116302> +/- 5.00m (speed -1.00 mps / course -1.00) @ 1/16/18, 12:15:32 AM Pacific Standard Time

╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager locationManager:didUpdateLocations:] Enabled: 1 | isMoving: 1 | df: 10.0m | age: 0.0s
╚═══════════════════════════════════════════════════════════
🔵-[TSLocationManager calculateMedianLocationAccuracy:] Median location accuracy: 5.0
🔵-[Settings incrementOdometer:] 4485751.8
ℹ️-[LocationDAO shrink:]_block_invoke maxRecordsToPersist: 1
✅-[TSLocationManager persistLocation:]_block_invoke INSERT: 84812B8F-5714-4367-97CC-F4CD0229C320

╔═══════════════════════════════════════════════════════════
║ -[TSHttpService flush:error:] 
╚═══════════════════════════════════════════════════════════
✅-[BackgroundTaskManager createBackgroundTask] 294
✅-[TSHttpService schedulePost:] LOCKED: 84812B8F-5714-4367-97CC-F4CD0229C320
🔵-[TSHttpService doPost:] HTTP Response: 200
✅-[TSHttpService post:error:] DESTROY: 84812B8F-5714-4367-97CC-F4CD0229C320

╔═══════════════════════════════════════════════════════════
║ -[TSHttpService finish:error:] Success: 1
╚═══════════════════════════════════════════════════════════
✅-[BackgroundTaskManager stopBackgroundTask:]_block_invoke 294 OF (
    294
)

📍<+43.98601004,-72.51314802> +/- 5.00m (speed -1.00 mps / course -1.00) @ 1/16/18, 12:15:33 AM Pacific Standard Time

╔═══════════════════════════════════════════════════════════
║ -[TSLocationManager locationManager:didUpdateLocations:] Enabled: 1 | isMoving: 1 | df: 10.0m | age: 0.0s
╚═══════════════════════════════════════════════════════════
🔵-[TSLocationManager calculateMedianLocationAccuracy:] Median location accuracy: 5.0
🔵-[Settings incrementOdometer:] 4495751.8
ℹ️-[LocationDAO shrink:]_block_invoke maxRecordsToPersist: 1
✅-[TSLocationManager persistLocation:]_block_invoke INSERT: 34C7E2F2-0A12-484F-97FA-0DD2CD29A048

╔═══════════════════════════════════════════════════════════
║ -[TSHttpService flush:error:] 
╚═══════════════════════════════════════════════════════════
✅-[BackgroundTaskManager createBackgroundTask] 295
✅-[TSHttpService schedulePost:] LOCKED: 34C7E2F2-0A12-484F-97FA-0DD2CD29A048
🔵-[TSHttpService doPost:] HTTP Response: 200
✅-[TSHttpService post:error:] DESTROY: 34C7E2F2-0A12-484F-97FA-0DD2CD29A048

╔═══════════════════════════════════════════════════════════
║ -[TSHttpService finish:error:] Success: 1
╚═══════════════════════════════════════════════════════════
✅-[BackgroundTaskManager stopBackgroundTask:]_block_invoke 295 OF (
    295
)
```

But let me try with your changes to `maxRecordsToPersist > 0` perhaps that's the reason for the lack of console.log.

**Comment by christocracy** [comment]
It's not `#onHttp(callback)`, it's `on('http', callback)`

**Comment by pencilcheck** [comment]
I tried both, none of them print anything
```
    BackgroundGeolocation.on('http', function(response) {
      console.log("- HTTP success callback");
      var status = response.status;
      var responseText = response.responseText;
      var res = JSON.parse(responseText);  // <-- if your server returns JSON
      console.log("- HTTP success", status, res);
    }, function(response) {
      console.log("- HTTP failure callback");
      var status = response.status;
      var responseText = response.responseText;
      console.log("- HTTP failure: ", status, responseText);
    })
```

**Comment by pencilcheck** [comment]
Also, I thought `extras` would merge with the location data, it seems like it is not doing it. As I expected to send this information to the backend.
```
POST
{
  query: "",
  variables: {
    "input": {
      "ping": { ... } // this merge extras with location data formatted in locationTemplate
    }
  }
}
```

But I will open another issue for that.

**Comment by christocracy** [comment]
I'm not seeing any issues with `http` event in Simulator.
![](https://dl.dropboxusercontent.com/s/yxfw3zrbojspmxa/Screenshot%202018-01-16%2008.20.47.png?dl=1)

**Comment by christocracy** [comment]
The plugin's json-templating is *very* simple.  It's not *intelligently* applying the `#extras` (ie: recursively) to the rendered template.  If you define `#extras` with the same structure as your `locationTemplate`, you're going to overwrite your location data.

**Comment by pencilcheck** [comment]
@christocracy I don't know why the callbacks weren't called in my end then. I am using the latest version and I setup based on the guide (I think that section might've been outdated since I followed the example code snippet in the guide for that section, you should update the README)

@christocracy well, I need to have values to be in the nested structure, the current attributes/methods provided can't satisfy my needs so I have to go with that hacky solutions I made by changing the templates every time directly.

**Comment by christocracy** [comment]
Why is it not possible to change your server code to consume a different schema?
